### DISCUSSION TITLE
Third-party adapters to the nextauthjs organization

### DISCUSSION THREAD

**Discussion by balazsorban44** [nextauthjs/next-auth#1161] [Ideas]
Currently, we would like to have a minimal number of built-in adapters for easier maintenance.
At the same time, we would like to encourage the community to create their own, custom adapter to support whatever database they would like to.

We could create a new repository to hold those third-party adapters, to give them a more official feeling.

As I see it, on NPM, they could be published in two ways:

~- `@next-auth/adapters` Including all the adapters in one package, each of them exported as their own module~ Ruled out.
- `@next-auth/{adapter-name}-adapter` each adapter is its own package, and the user only pulls down what they need

Ref: https://github.com/nextauthjs/next-auth/issues/1146

**Comment by pbteja1998** [comment]
Since Fauna is the only custom adapter we have at the moment, I think either of these two should be fine. I lean more towards second option mainly because it will be easy to maintain, especially since these are maintained by community and the people who use/maintain one custom adapter wouldn't use/maintain other custom adapter.

**Comment by ndom91** [comment]
Yeah I agree with @pbteja1998 and putting adapters in their own packages. Since DB adapters are mutually exclusive, a user would only ever be using one at a time. No need to download and install all the stuff for other databases as well.

**Comment by nyedidikeke** [comment]
> As I see it, on NPM, they could be published in two ways:
> 
> * `@next-auth/adapters` Including all the adapters in one package, each of them exported as their own module
> * `@next-auth/{adapter-name}-adapter` each adapter is its own package, and the user only pulls down what they need

The second option `@next-auth/{adapter-name}-adapter` is preferred for modularity, visibility and ease of maintenance.
Also, it will enable users save little bandwidth by downloading only the needed adapter(s).

**Reply by iaincollins** [reply]
Love this second option! Was thinking the same!

**Reply by balazsorban44** [reply]
Right! I am leaning towards something like this even with the built-in/default TypeORM adapter. Right now it will download `typeorm` to anyone (even when you don't even use a database), which is like 60% of the bundle size according to Bundlephobia. `typeorm` SHOULD be an optional dependency as well.

**Reply by pbteja1998** [reply]
Wow, didn't know that next-auth is more than 200 kb.

**Reply by balazsorban44** [reply]
Not that big of a deal since it's backend, but yeah.. We should make an effort to reduce it anyway. Especially if we want to introduce `openid-client`

**Comment by balazsorban44** [comment]
I made the repo https://github.com/nextauthjs/adapters public, so we can let people start streaming in with their custom adapters PRs!
Spread the word people!

**Comment by pbteja1998** [comment]
What's been decided?  Will there be separate repos for each adapter? or All of the adapters are in the same repo?

**Reply by balazsorban44** [reply]
I'm thinking one repo, the one I linked. Then one folder for each adapter, as well as one module/package per adapter on npm.

**Comment by ndom91** [comment]
Hey everyone, just wanted to summarize the current status of the adapter repo and figure out what steps need to be taken next.

So we have **(1) Prisma**  and **(2) TypeORM** merged and in. Both have some tests already as well.

We have PRs open for adding **(3) FaunaDB** and **(4) Firebase**. I think Fauna should be relatively easy to add as that was already in the main package at one point. @pbteja1998 if you need any help or have any questions, feel free to write in here. Regarding the Firebase adapter, that PRs mine - I'm just basing it off of the adapter a contributor sent in which he was successfully using his project. Primary work here is in "generalising" it, which shouldn't be too hard. 

We also talked to someone who has a **(5) DynamoDB** adapter on npm about contributing it upstream to this repo / set of packages. He said he was interested, but haven't heard anything since then (see: https://twitter.com/tgandrews/status/1355990745237348356)

I want to begin testing actually using the adapters from this repo in a branch of `next-auth`. What would you guys recommend as the best way to go about testing that? I was thinking about using our `next-auth-example` repo, taking a branch of `next-auth` with the adapters ripped out and the adapters from here `imported()` in. Does anyone have any better ideas, maybe just using the `next-auth` repo itself? I feel like I need an actual app using `next-auth` though to really test this (i.e. our example-app repo). 

Cheers!

CC: @balazsorban44 @iaincollins
